src/avl.py

```diff
@@ -0,0 +1,468 @@
+import json
+import logging
+
+from hypothesis import given
+import hypothesis.strategies as st
+
+
+
+class AVLTree():
+    """The interface to an AVL tree.
+
+    The tree itself is actually represented by AVLTreeNode instances, but because of
+    needing to be able to change which node is the root, we need a container class.
+
+    Args:
+      value: an initial value to insert into the tree.
+    """
+
+    def __init__(self, value=None):
+        self._root = AVLTreeNode(self)
+        if value is not None:
+            self._root.insert(value)
+
+        self.log = []
+
+    def insert(self, value):
+        """Inserts the vallue into the tree.
+
+        Args:
+          value: the value to insert into the tree
+        """
+
+        self.root.insert(value)
+
+    def write_log(self, file_name="tree_log.json"):
+        """Writes a log of operation description and tree pairs in JSON for visualization."""
+
+        with open(file_name, "w") as f:
+            f.write(json.dumps(self.log))
+
+    def __str__(self):
+        return str(self.root)
+
+    @property
+    def root(self):
+        return self._root
+
+    @root.setter
+    def root(self, value):
```

Comment:
what's the alternative? of course we can always directly set `._root`, but then we have to make sure we do the `._root.parent` setting everywhere too. this was the cause of a bug that caused me tremendous amounts of stress, so I'm disinclined to lose this benefit, but I'm totally open to suggestions that are as good as this approach in this regard.